src/library.js

```diff
@@ -3027,9 +3035,119 @@ LibraryManager.library = {
   msync: function(addr, len, flags) {
     // int msync(void *addr, size_t len, int flags);
     // http://pubs.opengroup.org/onlinepubs/009696799/functions/msync.html
-    // Pretend to succeed
+    // TODO: support sync'ing parts of allocations
+    var info = _mmap.mappings[addr];
+    if (!info) return 0;
+    if (len == info.num) {
+      var buffer = new Uint8Array(HEAPU8.buffer, addr, len);
+      return FS.msync(FS.getStream(info.fd), info, buffer, 0, len, flags);
+    }
+
     return 0;
   },
+  
+  // ==========================================================================
+  // emscripten_mmap.h
+  // ==========================================================================
+  mmap_read: function(heap, addr, addrOnHeap, type, index) {
+    var info = _mmap.mappings[addr];
+    // ?? How to handle this? 0 is a weird return value in a way...
+    if (!info) return 0;
+    return FS.mmap_read(FS.getStream(info.fd), info, type, index, heap[addrOnHeap + index]);
+  },
+  mmap_write: function(heap, addr, addrOnHeap, type, index, value) {
+    var info = _mmap.mappings[addr];
+    if (!info) return -1;
+    // Always write to heap too
+    heap[addrOnHeap + index] = value;
+    return FS.mmap_write(FS.getStream(info.fd), info, type, index, value);
+  },
+
+  mmap_read_char: function(addr, index) {
+    return _mmap_read(HEAP8, addr, addr, 'int8_t', index);
+  },
+  mmap_read_short: function(addr, index) {
+    return _mmap_read(HEAP16, addr, addr >> 1, 'int16_t', index);
```

Comment:
In emscripten land, short is always 2 bytes (if I'm not mistaken). But this is no guarantee on the host system where node.js runs native. Could be that it means something else. Therefore normalize short/long/etc. always to normal int16_t/int32_t.